### ISSUE TITLE
[dev.icinga.com #13889] Override Servicevars

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icingaweb2-module-director#696]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/13889

**Created by Finn on 2017-01-05 08:23:15 +00:00**

Assignee: _(none)_
Status: _New_ 
Target Version: _(none)_
Last Update: _2017-01-05 09:27:15 +00:00 (in Redmine)_

---

I have an issue when deploying Services with a costumvar.
In the Attachment you can see, that the costumvar is configured in the host but not applied to the service.

**Attachments**

* [vergleich.png](https://dev.icinga.com/attachments/download/4005/vergleich.png) Finn - _2017-01-05 08:20:55 +00:00_

**Comment by Thomas-Gelf** [comment]
Well, I guess the service has then set those variables (or inherited them) and therefore wins over the host vars?

**Comment by itsopensource** [comment]
Hello Guys, 
first of all thank you very much for the growing Director! Its getting more and more bigger. 
In our environment (latest icinga and latest Director-master 05a5f9f) we also have a problem with custom vars in services.
**The Setting**: We want to check the ntp time offset via ssh, where the most vars are definded in the Service.  

Our service looks like this:

```
template Service "service_linux-ntp" {
    import "servicetmpl_ssh-linux"
    check_command = "cmd_by_ssh"
    vars.by_ssh_command = "/usr/local/icinga/libexec/check_ntp_time -w $vars.ntp_warning$ -c $vars.ntp_critical$ $vars.ntp_args$"
    vars.by_ssh_logname = "icinga"
    vars.ntp_args = "-H timeserver1"
    vars.ntp_critical = "1"
    vars.ntp_warning = "0.5"
}

apply Service "Linux NTP" {
    import "service_linux-ntp"
    assign where host.name == "Linux1"
    vars.by_ssh_command = "/usr/local/icinga/libexec/check_ntp_time -w " + vars.ntp_warning + " -c " + vars.ntp_critical + " " + vars.ntp_args
    import DirectorOverrideTemplate
}
```

All works as designed. If we now want to change the Hostname of the NTP Host from "timeserver1" to "timeserver2" only for one host in the Director (_Hosts -> "Linux1" -> Services -> "Linux NTP" -> Change the Value of the field vars.ntp_args from "-H timeserver1" to "-H timeserver2"_ ), the rendered config looks like this:

```
object Host "Linux1" {
     import "hosttmpl_linux"
     display_name = "Linux1"
     address = "127.0.0.1"
     vars._override_servicevars = {
         "Linux Memory" = {
             memory_warning = "30"
         }
         "Linux NTP" = {
             ntp_args = "-H timeserver2"
         }
     }
}
```

**The Problem**: The executed Icinga2 Service shows the following vars:  "By Ssh Command" = "/usr/local/icinga/libexec/check_ntp_time -w 0.5 -c 1 -H timeserver1", the line below contains "Ntp Args" = "-H timeserver2" (see Screenshot)
![screenshot-icinga](https://cloud.githubusercontent.com/assets/25298537/22205679/3336ad24-e178-11e6-9642-0b438bee078a.png)
. 
The Command still tries the "timeserver1" host and the override of the value doesnt work.

```
<5> => get_service("Linux1","Linux NTP").last_check_result.command
[ "/usr/lib64/nagios/plugins/check_by_ssh", "-C", "/usr/local/icinga/libexec/check_ntp_time -w 0.5 -c 1 -H timeserver1", "-H", "Linux1", "-l", "icinga", "-t", "60" ]
```

Please help us to fix this. If you need more information or logs pls let us know. 
Thank you very very much ! :100:

**Comment by itsopensource** [comment]
Hello again, 
we found out something new:
under the "Services" tab for a single Host (where the Director renders the "object Service 'Custom NTP Service'..."  stuff) it is not possible to encapsulate a var into another var **when using the SSH command**, as shown in the example above. We guess the delivery from the director to icinga2 is not 100% correct. Because we are no really good in reading php code, we ask if anybody else have the same problem or perhaps a fix for that issue?

**Comment by Thomas-Gelf** [comment]
This should have been fixed with b475aa8
